origin_test.go

```
@@ -0,0 +1,382 @@
+
+func TestArtifactOrigin(t *testing.T) {
+	cases := []struct {
+		name string
+		raw  string
+		want string
+	}{
+		{name: "registry tarball", raw: "https://registry.npmjs.org/react/-/react-18.2.0.tgz", want: "https://registry.npmjs.org/react/-/react-18.2.0.tgz"},
+		{name: "checksum fragment is stripped", raw: "https://registry.npmjs.org/react/-/react-18.2.0.tgz#ceeba773e3e9d2b6f1a2b6b9f4f1cb2f9c2e1a55", want: "https://registry.npmjs.org/react/-/react-18.2.0.tgz"},
+		{name: "uppercase scheme is normalized", raw: "HTTPS://files.pythonhosted.org/packages/x/django-5.0.tar.gz", want: "https://files.pythonhosted.org/packages/x/django-5.0.tar.gz"},
+		{name: "signed link carrying a query", raw: "https://nexus.corp/repo/pkg.tgz?token=abc123"},
+		{name: "embedded credentials", raw: "https://user:s3cret@nexus.corp/repo/pkg.tgz"},
```

golangci-lint (2.12.2)

[high] G101: Potential hardcoded credentials: Password in URL

(gosec)